src/data/rat/defs.lean

```diff
@@ -461,17 +466,28 @@ instance : field ℚ :=
   mul_comm         := rat.mul_comm,
   mul_assoc        := rat.mul_assoc,
   left_distrib     := rat.mul_add,
-  right_distrib    := rat.add_mul,
+  right_distrib    := rat.add_mul }
+
+instance : comm_group_with_zero ℚ :=
+{ zero := 0,
+  one := 1,
+  mul := (*),
+  inv := has_inv.inv,
+  div := (/),
   exists_pair_ne   := ⟨0, 1, rat.zero_ne_one⟩,
-  mul_inv_cancel   := rat.mul_inv_cancel,
-  inv_zero         := rfl }
+  inv_zero := rfl,
+  mul_inv_cancel := rat.mul_inv_cancel,
+  mul_zero := mul_zero,
+  zero_mul := zero_mul,
+  .. rat.comm_ring }
+
+instance : is_domain ℚ :=
+{ .. rat.comm_group_with_zero,
+  .. (infer_instance : no_zero_divisors ℚ) }
 
 /- Extra instances to short-circuit type class resolution -/
```

Comment:
These extra instances are no longer going to go via `rat.field`, which is probably why you're hitting elaboration problems elsewhere.

My hunch would be that everything would actually behave better if you assembled these shortcuts first, building stronger structures out of weaker ones.

Reply:
There was only one real elaboration issue, in [counterexamples/pseudoelement.lean](https://github.com/leanprover-community/mathlib/pull/14893/files#diff-3c2f495ffe9b96910c43c53f664f803920d2135d04ea1df11d251694a5db9006), and that one already had a painful workaround. On the other hand, defining the weaker structures first means reproving results that are currently already deduced from others (see e.g. the `mul_zero` and `zero_mul` fields of the `add_comm_group_with_zero` instance). So I don't think it's worth it to redo all these definitions.

(Also I am not sure these "shortcut instances" are really worth it: see e.g. #1561)